Fix checkpointing and asynchronous helpers

jponge opened the pull request:
Do not let the asynchronous result test helpers create checkpoints

Fixes #5 
Fixes #31

jponge commented:
@dadoonet Can you please have a look?

dadoonet commented:
I built your branch and tried to modify my test but it's still unclear to me how this should work. I ended up writing:

```java
@ExtendWith(VertxExtension.class)
class VertxTest {

    private final Logger logger = LogManager.getLogger(this.getClass());

    @BeforeEach
    void deployVertice(Vertx vertx, VertxTestContext testContext) {
        vertx.deployVerticle(new FakeVertice(), testContext.succeeding(id -> {
            logger.info(" -> test webhook verticle deployed {}", id);
            logger.info("deployVertice {}", vertx.deploymentIDs());
            testContext.completeNow();
        }));
    }

    @AfterEach
    void stopVerticle(Vertx vertx, VertxTestContext testContext) {
        vertx.close(testContext.succeeding(f -> testContext.completeNow()));
    }

    @Test
    void testDeployVertice(Vertx vertx, VertxTestContext testContext) {
        testContext.verify(() -> {
            logger.info("testDeployVertice {}", vertx.deploymentIDs());
            assertFalse(vertx.deploymentIDs().isEmpty());
            testContext.completeNow();
        });
    }

    @Test
    void testPostValidate(Vertx vertx, VertxTestContext testContext) {
        // Create the WebClient to connect to our Vertx instance
        WebClient webClient = WebClient.create(vertx,
                new WebClientOptions()
                        .setDefaultHost("localhost")
                        .setDefaultPort(8080));

        HttpRequest<Buffer> request = webClient.get("/");
        request.send(testContext.succeeding(result -> testContext.verify(() -> {
            logger.fatal("CODE {}", result.statusCode());
            assertEquals(200, result.statusCode());

            logger.fatal("JSON {}", result.bodyAsString());
            logger.fatal("MESSAGE {}", result.bodyAsJsonObject().getString("message"));
            logger.fatal("JSON {}", result.bodyAsString());
        })));
    }


    static class FakeVertice extends AbstractVerticle {

        @Override
        public void start(Future<Void> futureVertx) {
            Router router = Router.router(vertx);
            router.route(HttpMethod.GET, "/")
                    .handler(routingContext -> {
                        // Write to the response and end it
                        HttpServerResponse response = routingContext.response();
                        response.setStatusCode(400);
                        String json = new JsonObject().put("message", "Vert.x is running!").encodePrettily();
                        response.putHeader("content-type", "application/json");
                        response.putHeader("content-length", "" + json.length());
                        response.write(json).end();
                    });

            vertx.createHttpServer().requestHandler(router::accept)
                    .listen(8080, ar -> {
                        if (ar.succeeded()) {
                            futureVertx.complete();
                        }
                        if (ar.failed()) {
                            futureVertx.fail("Can not start");
                        }
                    });
        }
    }
}
```

But this is still not working as I'd expect (it's not failing basically).

Could you tell how I should rewrite this test class?

jponge commented:
You need to end the test in `testPostValidate`. Also there is no need to cleanup the `vertx` instance when injected, this is being done automatically for you in the context. Last tip: you can use codecs with the web client, so you can directly get buffers, strings and json.

Here the full test, fixed and rewritten:

```java
package io.vertx.junit5;

import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.Vertx;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.client.HttpRequest;
import io.vertx.ext.web.client.WebClient;
import io.vertx.ext.web.client.WebClientOptions;
import io.vertx.ext.web.codec.BodyCodec;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

@ExtendWith(VertxExtension.class)
class DavidTest {

  private final Logger logger = LoggerFactory.getLogger(DavidTest.class);

  @BeforeEach
  void deployVertice(Vertx vertx, VertxTestContext testContext) {
    vertx.deployVerticle(new FakeVertice(), testContext.succeeding(id -> {
      logger.debug("Deployed!");
      testContext.completeNow();
    }));
  }

  @Test
  void testDeployVertice(Vertx vertx, VertxTestContext testContext) {
    testContext.verify(() -> {
      logger.info("Deployed ids: {}", vertx.deploymentIDs());
      assertFalse(vertx.deploymentIDs().isEmpty());
      testContext.completeNow();
    });
  }

  @Test
  void testPostValidate(Vertx vertx, VertxTestContext testContext) {
    WebClient webClient = WebClient.create(vertx,
      new WebClientOptions()
        .setDefaultHost("localhost")
        .setDefaultPort(8080));

    webClient
      .get("/")
      .as(BodyCodec.jsonObject())
      .send(testContext.succeeding(result -> testContext.verify(() -> {
        logger.debug("Response status code: {}", result.statusCode());
        assertEquals(200, result.statusCode());
        logger.debug("Response: {}", result.body().encodePrettily());
        testContext.completeNow();
      })));
  }


  static class FakeVertice extends AbstractVerticle {

    @Override
    public void start(Future<Void> futureVertx) {
      Router router = Router.router(vertx);
      router.route(HttpMethod.GET, "/")
        .handler(routingContext -> {
          // Write to the response and end it
          HttpServerResponse response = routingContext.response();
          response.setStatusCode(400);
          String json = new JsonObject().put("message", "Vert.x is running!").encodePrettily();
          response.putHeader("content-type", "application/json");
          response.putHeader("content-length", "" + json.length());
          response.write(json).end();
        });

      vertx.createHttpServer().requestHandler(router::accept)
        .listen(8080, ar -> {
          if (ar.succeeded()) {
            futureVertx.complete();
          }
          if (ar.failed()) {
            futureVertx.fail("Can not start");
          }
        });
    }
  }
}
```

And you can see it fail properly:

<img width="1440" alt="screen shot 2018-01-26 at 22 42 35" src="https://user-images.githubusercontent.com/25961/35462000-7e496150-02ea-11e8-9cad-9d08067d04ba.png">

jponge commented:
@dadoonet You may also explore checkpoints.

See:

```java
  @Test
  void testPostValidate(Vertx vertx, VertxTestContext testContext) {
    WebClient webClient = WebClient.create(vertx,
      new WebClientOptions()
        .setDefaultHost("localhost")
        .setDefaultPort(8080));

    Checkpoint httpRequestsCheckpoint = testContext.checkpoint(10);

    IntStream.range(0, 10).forEach(i -> {
      webClient
        .get("/")
        .as(BodyCodec.jsonObject())
        .send(testContext.succeeding(result -> testContext.verify(() -> {
          logger.debug("Response #{} status code: {}", i, result.statusCode());
          assertEquals(400, result.statusCode());
          logger.debug("Response #{} payload: {}", i, result.body().encodePrettily());
          httpRequestsCheckpoint.flag();
        })));
    });
  }
```

<img width="1440" alt="screen shot 2018-01-26 at 22 49 48" src="https://user-images.githubusercontent.com/25961/35462189-3ad5cb92-02eb-11e8-90dd-5d59056d1476.png">

dadoonet commented:
Ok. So for whatever reason the code I was using was not coming from your branch apparently although I built it locally.
I'd say that the PR is giving much better results now.

BUT (and it might be caused by another issue), I can't get the Json Body response from the test. It's always `null`.

```java
    @BeforeEach
    void deployVertice(Vertx vertx, VertxTestContext testContext) {
        vertx.deployVerticle(new FakeVertice(), testContext.succeeding(id -> {
            logger.debug("Deployed!");
            testContext.completeNow();
        }));
    }

    @Test
    void testPostValidate(Vertx vertx, VertxTestContext testContext) {
        WebClient webClient = WebClient.create(vertx,
                new WebClientOptions()
                        .setDefaultHost("localhost")
                        .setDefaultPort(8080));

        webClient
                .get("/")
                .as(BodyCodec.jsonObject())
                .send(testContext.succeeding(result -> testContext.verify(() -> {
                    logger.debug("Response status code: {}", result.statusCode());
                    assertEquals(200, result.statusCode());
                    String body = result.bodyAsString();
                    logger.debug("Response: {}", body);
                    assertNotNull(body);
                    assertEquals("foo", body);
                })));
    }


    static class FakeVertice extends AbstractVerticle {

        @Override
        public void start(Future<Void> futureVertx) {
            Router router = Router.router(vertx);
            router.route(HttpMethod.GET, "/")
                    .handler(routingContext -> {
                        // Write to the response and end it
                        HttpServerResponse response = routingContext.response();
                        response.setStatusCode(200);
                        String json = new JsonObject().put("message", "Vert.x is running!").encodePrettily();
                        response.putHeader("content-type", "application/json");
                        response.putHeader("content-length", "" + json.length());
                        response.write(json).end();
                    });

            vertx.createHttpServer().requestHandler(router::accept)
                    .listen(8080, ar -> {
                        if (ar.succeeded()) {
                            futureVertx.complete();
                        }
                        if (ar.failed()) {
                            futureVertx.fail("Can not start");
                        }
                    });
        }
    }
```

Any idea?

jponge commented:
There is a codec to have the response body as json, but you are using the
method to get it as a string. Try just using body() as I did in my example.

Le 27 janv. 2018 12:23, "David Pilato" <notifications@github.com> a écrit :

Ok. So for whatever reason the code I was using was not coming from your
branch apparently although I built it locally.
I'd say that the PR is giving much better results now.

BUT (and it might be caused by another issue), I can't get the Json Body
response from the test. It's always null.

    @BeforeEach
    void deployVertice(Vertx vertx, VertxTestContext testContext) {
        vertx.deployVerticle(new FakeVertice(), testContext.succeeding(id ->
 {
            logger.debug("Deployed!");
            testContext.completeNow();
        }));
    }


@Test
    void testPostValidate(Vertx vertx, VertxTestContext
 testContext) {

WebClient webClient = WebClient.create(vertx,
                new WebClientOptions()
                        .setDefaultHost("localhost")
                        .setDefaultPort(8080
));

        webClient
                .get("/")
                .as(BodyCodec.jsonObject())
                .send(testContext.succeeding(result ->
testContext.verify(() -> {
                    logger.debug("Response status code: {}",
result.statusCode());
                    assertEquals(200, result.statusCode());
                    String body = result.bodyAsString();
                    logger.debug("Response: {}", body);
                    assertNotNull(body);
                    assertEquals("foo", body);
                })));
    }



static class FakeVertice extends AbstractVerticle {

        @Override
        public void start(Future<Void> futureVertx) {
            Router router = Router.router(vertx);
            router.route(HttpMethod.GET, "/")
                    .handler(routingContext -> {
                        // Write to the response and end it
                        HttpServerResponse response = routingContext.
response();
                        response.setStatusCode(200);

String json = new JsonObject().put("message", "Vert.x is
running!").encodePrettily();
                        response.putHeader("content-type", "application/json");
                        response.putHeader("content-length", "" +
json.length());
                        response.write(json).end();
                    });

            vertx.createHttpServer().requestHandler(router::accept)
                    .listen(8080, ar -> {
                        if (ar.succeeded()) {
                            futureVertx.complete();
                        }
                        if (ar.failed()) {
                            futureVertx.fail("Can not start");
                        }
                    });
        }
    }

Any idea?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/vert-x3/vertx-junit5/pull/32#issuecomment-360978338>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AABladVW_GmLmIvg-h_j9XFtLbT9w3j_ks5tOwc4gaJpZM4RuXYK>
.

jponge commented:
That being said if you get null with the codec it may be worth reporting
the behavior to vertx-web-client.

Le 27 janv. 2018 15:32, "Julien Ponge" <julien.ponge@gmail.com> a écrit :

> There is a codec to have the response body as json, but you are using the
> method to get it as a string. Try just using body() as I did in my example.
>
> Le 27 janv. 2018 12:23, "David Pilato" <notifications@github.com> a
> écrit :
>
> Ok. So for whatever reason the code I was using was not coming from your
> branch apparently although I built it locally.
> I'd say that the PR is giving much better results now.
>
> BUT (and it might be caused by another issue), I can't get the Json Body
> response from the test. It's always null.
>
>     @BeforeEach
>     void deployVertice(Vertx vertx, VertxTestContext testContext) {
>         vertx.deployVerticle(new FakeVertice(), testContext.succeeding(id ->
>  {
>             logger.debug("Deployed!");
>             testContext.completeNow();
>         }));
>     }
>
>
> @Test
>     void testPostValidate(Vertx vertx, VertxTestContext
>  testContext) {
>
> WebClient webClient = WebClient.create(vertx,
>                 new WebClientOptions()
>                         .setDefaultHost("localhost")
>                         .setDefaultPort(8080
> ));
>
>         webClient
>                 .get("/")
>                 .as(BodyCodec.jsonObject())
>                 .send(testContext.succeeding(result -> testContext.verify(() -> {
>                     logger.debug("Response status code: {}", result.statusCode());
>                     assertEquals(200, result.statusCode());
>                     String body = result.bodyAsString();
>                     logger.debug("Response: {}", body);
>                     assertNotNull(body);
>                     assertEquals("foo", body);
>                 })));
>     }
>
>
>
> static class FakeVertice extends AbstractVerticle {
>
>         @Override
>         public void start(Future<Void> futureVertx) {
>             Router router = Router.router(vertx);
>             router.route(HttpMethod.GET, "/")
>                     .handler(routingContext -> {
>                         // Write to the response and end it
>                         HttpServerResponse response = routingContext.
> response();
>                         response.setStatusCode(200);
>
> String json = new JsonObject().put("message", "Vert.x is running!").encodePrettily();
>                         response.putHeader("content-type", "application/json");
>                         response.putHeader("content-length", "" + json.length());
>                         response.write(json).end();
>                     });
>
>             vertx.createHttpServer().requestHandler(router::accept)
>                     .listen(8080, ar -> {
>                         if (ar.succeeded()) {
>                             futureVertx.complete();
>                         }
>                         if (ar.failed()) {
>                             futureVertx.fail("Can not start");
>                         }
>                     });
>         }
>     }
>
> Any idea?
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vert-x3/vertx-junit5/pull/32#issuecomment-360978338>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABladVW_GmLmIvg-h_j9XFtLbT9w3j_ks5tOwc4gaJpZM4RuXYK>
> .
>
>
>

jponge commented:
@dadoonet Are we good David?

dadoonet commented:
I did not have time to check your last answers but I'm planning to do it later this week.

But as I said:

> I'd say that the PR is giving much better results now.

So it looks good to me.
